### PR TITLE
Adding Contrail SDN support for OpenShift

### PR DESCRIPTION
**Pull request by savithruml**
RE: #5623

### CODE REVIEW

File: roles/openshift_common/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,83 @@
+---
```
Comment (mtnbikenc) [inline roles/openshift_common/tasks/main.yml]: This file should be removed from the PR.

File: roles/contrail_node/meta/main.yml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  min_ansible_version: 2.2
+  platforms:
+  - name: EL
+    versions:
+    - 7
+  categories:
+  - cloud
+  - system
+dependencies:
+- role: openshift_facts
```
Comment (mtnbikenc) [inline roles/contrail_node/meta/main.yml]: I do not believe either of these dependencies are needed.

File: roles/contrail_master/meta/main.yml [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  min_ansible_version: 2.2
+  platforms:
+  - name: EL
+    versions:
+    - 7
+  categories:
+  - cloud
+  - system
+dependencies:
+- role: openshift_facts
```
Comment (mtnbikenc) [inline roles/contrail_master/meta/main.yml]: I do not believe either of these dependencies are needed.

File: roles/contrail_master/templates/contrail-installer.j2 [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+    sandesh_ssl_enable = False
+    enable_config_service = True
+    enable_control_service = True
+    enable_webui_service = True
+    introspect_ssl_enable = False
+    config_nodes = {{ groups.oo_first_master.0 }} 
+    controller_nodes = {{ groups.oo_first_master.0 }}
+    analytics_nodes = {{ groups.oo_first_master.0 }}
+    analyticsdb_nodes = {{ groups.oo_first_master.0 }}
+    analyticsdb_minimum_diskgb = {{ analyticsdb_min_diskgb }}
```
Comment (mtnbikenc) [inline roles/contrail_master/templates/contrail-installer.j2]: There are various vars in use in the role and this template which should be documented in a README.md for the role.

File: roles/contrail_master/tasks/main.yaml [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    tag: "{{ contrail_version }}"
+    timeout: 300
+    state: present
+    load_path: "/tmp/contrail-kube-manager-{{ os_release }}-{{ contrail_version }}.tar.gz"
+  when: docker_registry is not defined and inventory_hostname == groups.oo_first_master.0
+
+- name: Create Contrail single YAML file
+  become: yes
+  template: src=contrail-installer.j2 dest=/tmp/contrail-installer.yaml owner=root mode=0644
+  when: inventory_hostname == groups.oo_first_master.0
```
Comment (ingvagabund) [inline roles/contrail_master/tasks/main.yaml]: Does it make sense to run this task if `docker_registry is not defined` holds?
Please, use the multiline line form for the module parameters, i.e.:
```yaml
template:
  src: contrail-installer.j2
  dest: /tmp/contrail-installer.yaml
  owner: root
  mode: 0644
```

Plus, 
```yaml
when: inventory_hostname == contrail_first_master
```

File: roles/contrail_master/templates/contrail-installer.j2 [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+    cloud_orchestrator = openshift 
+    sandesh_ssl_enable = False
+    enable_config_service = True
+    enable_control_service = True
+    enable_webui_service = True
+    introspect_ssl_enable = False
+    config_nodes = {{ groups.oo_first_master.0 }} 
+    controller_nodes = {{ groups.oo_first_master.0 }}
+    analytics_nodes = {{ groups.oo_first_master.0 }}
+    analyticsdb_nodes = {{ groups.oo_first_master.0 }}
```
Comment (ingvagabund) [inline roles/contrail_master/templates/contrail-installer.j2]: `s/groups.oo_first_master.0/contrail_first_master/g`

File: roles/contrail_node/meta/main.yml [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+- role: openshift_master_facts
```
Comment (ingvagabund) [inline roles/contrail_node/meta/main.yml]: Why master facts? Not node facts?

File: roles/contrail_node/tasks/main.yaml [hunk 8]
```diff
@@ -0,0 +1,47 @@
+---
+- name: Untar Contrail tar-file
+  unarchive:
+    src: "{{ contrail_docker_images_path }}/contrail-kubernetes-docker-images_{{ contrail_version }}.tgz"
+    dest: "/tmp"
+  when: docker_registry is not defined
```
Comment (ingvagabund) [inline roles/contrail_node/tasks/main.yaml]: Can you put all tasks with the `docker_registry is not defined` under a block with a single condition?

File: roles/contrail_node/tasks/main.yaml [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+  docker_image:
+    name: contrail-k8s-agent
+    tag: "{{ contrail_version }}"
+    timeout: 300
+    load_path: "/tmp/contrail-kubernetes-agent-{{ os_release }}-{{ contrail_version }}.tar.gz"
+  when: docker_registry is not defined
+
+- name: Add contrail service account to privileged scc
+  shell: oadm policy add-scc-to-user privileged system:serviceaccount:kube-system:contrail
+  delegate_to: "{{ groups.oo_first_master.0 }}"
```
Comment (ingvagabund) [inline roles/contrail_node/tasks/main.yaml]: `contrail_first_master`

File: roles/contrail_node/tasks/main.yaml [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+- name: Load Contrail-Kubernetes-Agent image
+  docker_image:
+    name: contrail-k8s-agent
+    tag: "{{ contrail_version }}"
+    timeout: 300
+    load_path: "/tmp/contrail-kubernetes-agent-{{ os_release }}-{{ contrail_version }}.tar.gz"
+  when: docker_registry is not defined
+
+- name: Add contrail service account to privileged scc
+  shell: oadm policy add-scc-to-user privileged system:serviceaccount:kube-system:contrail
```
Comment (ingvagabund) [inline roles/contrail_node/tasks/main.yaml]: @kwoodson do we have a module for it?

File: roles/contrail_node/tasks/main.yaml [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+  delegate_to: "{{ groups.oo_first_master.0 }}"
+  when: inventory_hostname == groups.oo_first_master.0
+
+- name: Add the daemonset-controller service account to privileged scc
+  shell: oc adm policy add-scc-to-user privileged system:serviceaccount:kube-system:daemon-set-controller
```
Comment (ingvagabund) [inline roles/contrail_node/tasks/main.yaml]: @kwoodson do we have a module for it? And the others below?

File: roles/contrail_node/tasks/main.yaml [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    name: contrail-k8s-agent
+    tag: "{{ contrail_version }}"
+    timeout: 300
+    load_path: "/tmp/contrail-kubernetes-agent-{{ os_release }}-{{ contrail_version }}.tar.gz"
+  when: docker_registry is not defined
+
+- name: Add contrail service account to privileged scc
+  shell: oadm policy add-scc-to-user privileged system:serviceaccount:kube-system:contrail
+  delegate_to: "{{ groups.oo_first_master.0 }}"
+  when: inventory_hostname == groups.oo_first_master.0
```
Comment (ingvagabund) [inline roles/contrail_node/tasks/main.yaml]: Can you put all tasks with the `inventory_hostname == contrail_first_master` under a block with a single condition?

**Comment by ingvagabund** [comment]
Can you get rid of `Merge pull request #1 from openshift/master ` commit and squash?

File: roles/openshift_common/tasks/main.yml (hunk 1 above)
Comment (savithruml) [reply roles/openshift_common/tasks/main.yml]: @mtnbikenc can you please point where "os_use_contrail" variable is used if I remove this file?

File: roles/contrail_master/tasks/main.yaml [hunk 13]
```diff
@@ -0,0 +1,47 @@
+---
+- name: Untar Contrail tar-file
+  unarchive:
+    src: "{{ contrail_docker_images_path }}/contrail-kubernetes-docker-images_{{ contrail_version }}.tgz"
+    dest: "/tmp"
+  when: docker_registry is not defined and inventory_hostname == groups.oo_first_master.0
```
[the remark this answers is not in the document]
Comment (savithruml) [reply roles/contrail_master/tasks/main.yaml]: @ingvagabund can you please provide an example?

[the remark this answers is not in the document]
Comment (ingvagabund) [reply roles/contrail_master/tasks/main.yaml]:
```yaml
...
  - role: contrail_master
    contrail_first_master: "{{ groups.oo_first_master.0 }}"
    when: openshift_use_contrail | default(false) | bool
```

File: roles/contrail_node/tasks/main.yaml [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+      tag: "{{ contrail_version }}"
+      timeout: 300
+      load_path: "/tmp/contrail-kubernetes-agent-{{ os_release }}-{{ contrail_version }}.tar.gz"
+
+- when:
+  - inventory_hostname == groups.oo_first_master.0
+
+  block:
+  - name: Add contrail service account to privileged scc
+    shell: oadm policy add-scc-to-user privileged system:serviceaccount:kube-system:contrail
```
Comment (michaelgugino) [inline roles/contrail_node/tasks/main.yaml]: Do we need shell here, or can we use command?

File: roles/contrail_node/tasks/main.yaml [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+    delegate_to: "{{ groups.oo_first_master.0 }}"
+
+  - name: Add the daemonset-controller service account to privileged scc
+    shell: oc adm policy add-scc-to-user privileged system:serviceaccount:kube-system:daemon-set-controller
```
Comment (michaelgugino) [inline roles/contrail_node/tasks/main.yaml]: Do we need shell here, or can we use command?

I think there's also a plugin for oc_adm that might be a better fit.

Comment (michaelgugino) [reply roles/contrail_node/tasks/main.yaml]: @kwoodson

File: roles/contrail_node/tasks/main.yaml [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+    delegate_to: "{{ groups.oo_first_master.0 }}"
+
+  - name: Label master nodes with opencontrail.org/controller=true
+    shell: oc label nodes {{ openshift.common.hostname }} opencontrail.org/controller=true --overwrite=true
```
Comment (michaelgugino) [inline roles/contrail_node/tasks/main.yaml]: See openshift_manage_node role for example on using a plugin to do this.

File: roles/contrail_node/tasks/main.yaml [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+    delegate_to: "{{ groups.oo_first_master.0 }}"
+
+  - name: Make master schedulable
+    shell: oadm manage-node {{ openshift.common.hostname }} --schedulable
```
Comment (michaelgugino) [inline roles/contrail_node/tasks/main.yaml]: Does this service *have* to run on a master node exclusively?

We already have the concept of infra-nodes elsewhere in the project.  We'll probably need to discuss how best to accommodate this step.

File: roles/contrail_node/tasks/main.yaml [hunk 18]
```diff
@@ -0,0 +1,47 @@
+---
+- when:
+  - docker_registry is not defined
```
Comment (michaelgugino) [inline roles/contrail_node/tasks/main.yaml]: What is the purpose of this when statement?  What docker_registry?  Where would it be defined?  Can you provide some insight into this?

File: roles/contrail_node/tasks/main.yaml [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+    unarchive:
+      src: "{{ contrail_docker_images_path }}/contrail-kubernetes-docker-images_{{ contrail_version }}.tgz"
+      dest: "/tmp"
+
+  - name: Load Contrail-vRouter-Agent image
+    docker_image:
+      name: contrail-agent
+      tag: "{{ contrail_version }}"
+      timeout: 300
+      load_path: "/tmp/contrail-agent-{{ os_release }}-{{ contrail_version }}.tar.gz"
```
Comment (michaelgugino) [inline roles/contrail_node/tasks/main.yaml]: Let's namespace this variable better.

contrail_os_release would be more appropriate.

File: roles/contrail_node/tasks/main.yaml [hunk 20]
```diff
@@ -0,0 +1,47 @@
+---
+- when:
+  - docker_registry is not defined
+
+  block:
+  - name: Untar Contrail tar-file
+    unarchive:
+      src: "{{ contrail_docker_images_path }}/contrail-kubernetes-docker-images_{{ contrail_version }}.tgz"
```
Comment (michaelgugino) [inline roles/contrail_node/tasks/main.yaml]: This is a tarball containing two tarballs?

Does this match some vendor documentation?  Where is this tarball sourced?  I would like to see a link in the README.md that links to appropriate upstream documentation.

File: playbooks/common/openshift-node/additional_config.yml [hunk 21]
```diff
@@ -50,3 +54,9 @@
   - role: contiv
     contiv_role: netplugin
     when: openshift_use_contiv | default(false) | bool
+    
+- name: Additional node config
```
Comment (michaelgugino) [inline playbooks/common/openshift-node/additional_config.yml]: Thank you for following the existing pattern here!

File: roles/contrail_master/tasks/main.yaml (hunk 13 above)
[the remark this answers is not in the document]
Comment (savithruml) [reply roles/contrail_master/tasks/main.yaml]: @ingvagabund thank you! If I want to run on all masters, should I use `"{{ groups.oo_masters_to_config }}"`

File: roles/contrail_node/meta/main.yml (hunk 7 above)
Comment (savithruml) [reply roles/contrail_node/meta/main.yml]: @ingvagabund will `openshift-node` playbook invoke node facts?

File: roles/contrail_node/tasks/main.yaml (hunk 17 above)
Comment (savithruml) [reply roles/contrail_node/tasks/main.yaml]: @michaelgugino yes, this service has to be run on master node

File: roles/contrail_node/tasks/main.yaml (hunk 16 above)
Comment (savithruml) [reply roles/contrail_node/tasks/main.yaml]: @michaelgugino in openshift_manage_node, does node mean both master+nodes?

File: roles/contrail_master/tasks/main.yaml [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+      state: present
+      load_path: "/tmp/contrail-analyticsdb-{{ contrail_os_release }}-{{ contrail_version }}.tar.gz"
+
+  - name: Load Contrail-Kube-Manager image
+    docker_image:
+      name: contrail-kube-manager
+      tag: "{{ contrail_version }}"
+      timeout: 300
+      state: present
+      load_path: "/tmp/contrail-kube-manager-{{ contrail_os_release }}-{{ contrail_version }}.tar.gz"
```
Comment (ingvagabund) [inline roles/contrail_master/tasks/main.yaml]: What about
```yaml
 - name: Load contrail master docker images
    docker_image:
      name: "{{ item }}"
      tag: "{{ contrail_version }}"
      timeout: 300
      state: present
      load_path: "/tmp/{{ item }}-{{ contrail_os_release }}-{{ contrail_version }}.tar.gz"
  with_items:
  - contrail-controller
  - contrail-analytics
  - contrail-analyticsdb
  - contrail-kube-manager
```
?

File: roles/contrail_node/tasks/main.yaml [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+      tag: "{{ contrail_version }}"
+      timeout: 300
+      load_path: "/tmp/contrail-agent-{{ contrail_os_release }}-{{ contrail_version }}.tar.gz"
+
+  - name: Load Contrail-Kubernetes-Agent image
+    docker_image:
+      name: contrail-k8s-agent
+      tag: "{{ contrail_version }}"
+      timeout: 300
+      load_path: "/tmp/contrail-kubernetes-agent-{{ contrail_os_release }}-{{ contrail_version }}.tar.gz"
```
Comment (ingvagabund) [inline roles/contrail_node/tasks/main.yaml]: What about
```yaml
- name: Load contrail node docker images
  docker_image:
    name: "{{ item }}"
    tag: "{{ contrail_version }}"
    timeout: 300
    load_path: "/tmp/{{ item }}-{{ contrail_os_release }}-{{ contrail_version }}.tar.gz"
  with_items:
  - contrail-agent
  - contrail-k8s-agent
```
?

File: roles/contrail_master/templates/contrail-installer.j2 (hunk 6 above)
Comment (ingvagabund) [reply roles/contrail_master/templates/contrail-installer.j2]: The substitution is still applicable

File: roles/contrail_node/tasks/main.yaml [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+  block:
+  - name: Add contrail service account to privileged scc
+    oc_adm_policy_user:
+      user: system:serviceaccount:kube-system:contrail
+      resource_kind: scc
+      resource_name: privileged
+      state: present
+    delegate_to: "{{ contrail_first_master }}"
+
+  - name: Add the daemonset-controller service account to privileged scc
```
Comment (ingvagabund) [inline roles/contrail_node/tasks/main.yaml]: Are there going to be only two service accounts? If the list will grow I would consider using the `with_items` construction.

File: roles/openshift_common/tasks/main.yml [hunk 25]
```diff
@@ -0,0 +1,83 @@
...
+      use_flannel: "{{ openshift_use_flannel | default(None) }}"
+      use_calico: "{{openshift_use_calico | default(None) }}"
+      use_nuage: "{{ openshift_use_nuage | default(None) }}"
+      use_contiv: "{{ openshift_use_contiv | default(None) }}"
+      use_contrail: "{{ openshift_use_contrail | default(None) }}"
+      use_manageiq: "{{ openshift_use_manageiq | default(None) }}"
+      data_dir: "{{ openshift_data_dir | default(None) }}"
+      use_dnsmasq: "{{ openshift_use_dnsmasq | default(None) }}"
+
+- name: Install the base package for versioning
```
Comment (ingvagabund) [inline roles/openshift_common/tasks/main.yml]: This task is already part of the `openshift_version` role.

File: roles/openshift_common/tasks/main.yml (hunk 1 above)
Comment (ingvagabund) [reply roles/openshift_common/tasks/main.yml]: @savithruml why do we need this file?

**Comment by ingvagabund** [comment]
@savithruml each commit is a small change, squashing all the commits will make it look more transparent

File: roles/contrail_node/tasks/main.yaml [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+  - name: Label master nodes with opencontrail.org/controller=true
+    command: oc label nodes {{ openshift.common.hostname }} opencontrail.org/controller=true --overwrite=true
+    delegate_to: "{{ contrail_first_master }}"
+
+  - name: Make master schedulable
+    command: oadm manage-node {{ openshift.common.hostname }} --schedulable
+    delegate_to: "{{ contrail_first_master }}"
+
+  - name: Launch the installer
+    command: oc create -f /tmp/contrail-installer.yaml
```
Comment (kwoodson) [inline roles/contrail_node/tasks/main.yaml]: suggest using oc_obj

File: roles/contrail_node/tasks/main.yaml [hunk 27]
```diff
@@ -0,0 +1,55 @@
...
+      resource_name: privileged
+      state: present
+    delegate_to: "{{ contrail_first_master }}"
+
+  - name: Label master nodes with opencontrail.org/controller=true
+    command: oc label nodes {{ openshift.common.hostname }} opencontrail.org/controller=true --overwrite=true
+    delegate_to: "{{ contrail_first_master }}"
+
+  - name: Make master schedulable
+    command: oadm manage-node {{ openshift.common.hostname }} --schedulable
```
Comment (kwoodson) [inline roles/contrail_node/tasks/main.yaml]: suggest using oc_adm_manage_node.

File: roles/contrail_node/tasks/main.yaml [hunk 28]
```diff
@@ -0,0 +1,55 @@
...
+  - name: Add the daemonset-controller service account to privileged scc
+    oc_adm_policy_user:
+      user: system:serviceaccount:kube-system:daemon-set-controller
+      resource_kind: scc
+      resource_name: privileged
+      state: present
+    delegate_to: "{{ contrail_first_master }}"
+
+  - name: Label master nodes with opencontrail.org/controller=true
+    command: oc label nodes {{ openshift.common.hostname }} opencontrail.org/controller=true --overwrite=true
```
Comment (kwoodson) [inline roles/contrail_node/tasks/main.yaml]: suggest using oc_label

File: roles/contrail_master/tasks/main.yaml [hunk 29]
```diff
@@ -0,0 +1,50 @@
+---
+- when:
+  - docker_registry is not defined
+  - inventory_hostname == contrail_first_master
+
+  block:
+  - name: Untar Contrail tar-file
```
Comment (snuxoll) [inline roles/contrail_master/tasks/main.yaml]: OpenContrail images are now available on Docker Hub, can we pull these if users don't want to download the RHEL7 tarballs from Juniper's website (particularly for Origin deployments which may be done on CentOS).

**Comment by vrutkovs** [comment]
Master branch is closed! A major refactor is ongoing in devel-40. Changes for 3.x should be made directly to the latest release branch they're relevant to and backported from there.

**Pull request closed without merging**